src/ui.rs

```diff
@@ -1292,6 +1333,37 @@ impl<'state, 'input> Recorder<'state, 'input> {
         }
     }
 
+    fn select_prev_same_kind(&self, keys: &[SelectionKey], index: Option<usize>) -> SelectionKey {
+        match index {
+            None => self.first_selection_key(),
+            Some(index) => {
+                let current = keys[index];
+                for key in keys[index..].iter().chain(keys[..index].iter()).rev() {
+                    if std::mem::discriminant(key) == std::mem::discriminant(&current) {
+                        return *key;
+                    }
+                }
+                panic!("Unreachable. The search will loop around and return the current key.");
+            }
+        }
+    }
+
+    fn select_next_same_kind(&self, keys: &[SelectionKey], index: Option<usize>) -> SelectionKey {
+        match index {
+            None => self.first_selection_key(),
+            Some(index) => {
+                let current = keys[index];
+                let start_at = (index + 1) % keys.len();
+                for key in keys[start_at..].iter().chain(keys[..start_at].iter()) {
```

Comment:
suggestion: Index management can be tricky/have edge-cases. Let's extract the loop-around iterator into a single function to reduce the chance that one of these selection functions has a bug that the other doesn't have.

- How familiar are you with Rust? I don't know how much detail to write.
- In this situation, to extract the helper function, you'll likely want to return `impl Iterator<Item = ...>` rather than a  concrete iterator type.
- The indexing behavior is fundamentally different with regards to whether it includes the current index. Let's unify it by always excluding the current index (then `rev` should work as expected on the returned iterator).
  - The not-found case will therefore default to the current item instead of panicking.
- I'm guessing it'll then be straightforward to turn the `for` loop into a pipeline that uses `find` (or `filter` or `filter_map` or `find_map`) followed by `unwrap_or`/`unwrap_or_else`.

Reply:
Ok I did the following:

1. I restored the vim keybindings.
2. I still added the ctrl-up and ctrl-down scroll bindings
3. I added the ctrl modifier to the old page-up and page-down movement keys to make room for this new "next section of the same kind" feature

As far as the code, I ended up drying out this function completely so that both prev and next call the same function with a different boolean value. The end result isn't exactly the way you described it, but it accomplishes the same thing. Please take a look :)